### DISCUSSION TITLE
ACMultiInterface kernel: kappa as a function of a variable?

### DISCUSSION THREAD

**Discussion by scottedmuller-pnnl** [idaholab/moose#18299] [Q&A Modules: Phase field] [answered]
Hi everyone,

I'm trying to implement the ACMultiInterface kernel in a 3-phase system and I would like the kappa terms to be functions of a separate order parameter (a different one than the "eta" order parameters from the interface). If a kappa term is defined as a DerivativeParsedMaterial as a function of theta (for example), will the changing value of kappa with theta be reflected in the ACMultiInterface kernel? And more importantly, will the ACMultiInterface kernel include the gradient of theta as it computes the divergence of the interface terms? If using a kappa term as a function of another variable is not possible, is there a workaround to get an equivalent effect?

Thanks,
Scott

**Comment by SudiptaBiswas** [comment]
It's not clear to me what you are trying to implement here. Could you please elaborate on the model/formulation? Kernels associated with basic phase-field equations terms are provided here, https://mooseframework.inl.gov/modules/phase_field/Phase_Field_Equations.html 
You can also check out `ACKappaFunction`  which considers the derivative of the kappa term.

**Reply by scottedmuller-pnnl** [reply]
Thank you for the response. I'm trying to implement a non-isothermal system with three distinct non-conserved phases, each of which has its own order parameter (representing the phase volume fraction),
#### Variables:
    Order parameters: phi_1, phi_2, phi_3
    Temperature: T
For brevity, I'll use **phi_q** to reference the order parameters in this outline. The free energy is a combination of thermodynamic terms (heat capacity and latent heat with phase changes) and coupled interface terms that may be different between any two phases,
#### Free Energy:
    f(phi_q,T) = f_bulk(phi_q,T) + f_interface(phi_q,grad(phi_q))
    f_bulk(phi_q,T) = h(phi_1)*f_1(T) + h(phi_2)*f_2(T) + h(phi_3)*f_3(T)
    f_interface(phi_q,grad(phi_q)) = f_dw(phi_q) + f_grad(phi_q,grad(phi_q))
I felt the [ACMultiInterface](https://mooseframework.inl.gov/source/kernels/ACMultiInterface.html) kernel was most appropriate for f_grad as it handles the complex cross terms between each pair of order parameters and their gradients. I use a [CrossTermBarrierFunctionMaterial](https://mooseframework.inl.gov/source/materials/CrossTermBarrierFunctionMaterial.html) for the double-well term **f_dw**. The evolution equations for the order parameters are then straightforward,
#### Evolution equation for order parameter *q*:
    dphi_q/dt + L*df/dphi_q + penalty = 0 ;  q={1,2,3}
The challenge comes in deriving an evolution equation for temperature. The internal energy can be derived as
#### Internal energy:
    e(phi_q,T) = f(phi_q,T) - T*df/dT = e_bulk(phi_q,T) + f_interface(phi_q,grad(phi_q))
Note that the interface term is unchanged in this transformation because **df_interface/dT = 0**. The time evolution for temperature, with thermal conductivity **k**, can be expressed as,
#### Evolution equation for T:
    de/dT*dT/dt + sum(de/dphi_q*dphi_q/dt + de/d(grad(phi_q))*d(grad(phi_q))/dt,q={1,2,3}) - div(k(phi_q)*grad(T)) = 0
I'm not positive that the **grad(phi_q)** terms needs to be included here, but I don't feel that I can discard them out of hand. Even if I could, the **de/dphi_q** terms are themselves quite complicated as they must include portions of the cross terms from **f_interface**. It may be that the effect of these terms is negligible enough that they can safely be discarded, but I would have to actually set it up in order to verify it, which is what I hope to avoid.

To dramatically simplify this problem I could recast **f_interface** as being a linear function of T. The internal energy would then simply be **e(phi_q,T) = e_bulk(phi_q,T)** and the evolution equation for T would become
#### Recast internal energy and evolution equation for T:
    de_bulk/dT*dT/dt + sum(de_bulk/dphi_q*dphi_q/dt,q={1,2,3}) - div(k(phi_q)*grad(T)) = 0
Here the **de_bulk/dphi_q** terms are very straightforward. This modified formulation seems mathematically complete to me, and much less taxing computationally. I had hoped to implement it by making the kappa term in the ACMultiInterface kernel a simple linear function of T, but it doesn't seem to be a built-in option, hence my question.

The [ACKappaFunction](https://mooseframework.inl.gov/source/kernels/ACKappaFunction.html) you suggested looks promising. I suppose I would use a different ACKappaFunction for each kappa value for a given order parameter. Does that sound right? Looking at ACMultiInterface, do you have any reason to think it can't be used with multiple kappa values each defined by their own ACKappaFunction? Please let me know if you have any other ideas or suggestions. Also, please let me know if you see a problem with my formulation or approach in general. Thanks again for your help!

-Scott

**Comment by dschwen** [comment, marked as answer]
In the 4th equation on https://mooseframework.inl.gov/moose/source/kernels/ACMultiInterface.html you can see that we pulled out Kappa, assuming that it is constant (i.e. has a vanishing gradient). This is consistent with the code in `modules/phase_field/src/kernels/ACMultiInterface.C`, where no derivatives of the kappas are used. 

For your application you'd have to redo the derivation and implement the resulting weak form containing the dkappa/dvar derivatives for all variables your kappas depend on (to construct grad(kappa) = sum(grad(var)*dkappa/dvar).

**Reply by scottedmuller-pnnl** [reply]
That makes sense, and answers my original question. Thanks!

I'm probably overthinking it. The literature I've reviewed so far doesn't bother including interfacial energy terms with their internal energy when deriving temperature evolution equations for non-isothermal systems. Most seem to treat internal energy as a bulk quantity. I'm not fully convinced and would love to see a justification one way or another. I'd be grateful if you have any other advice related to my problem, or any related literature that deals with it more explicitly. 

-Scott
